adding unicycler 0.5.0

erinyoung opened the pull request:
- [X] This comment contains a description of what is in the pull request.

This is for the latest and greatest version of unicycler, which has gone through a lot of changes

<!-- If this PR is to add a new version of a tool that already has a dockerfile -->
- [X] Build your own docker image using a Dockerfile
  - [X] Directory structure should be name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. spades/3.12.0/Dockerfile)
  - [X] Includes the recommended LABELS
- [X] (Optional) Dockerfile is built with best practices and has been approved by a linter (such as https://hadolint.github.io/hadolint/)
- [X] Edit main README.md
- [X] Ensure tool is listed in Program_Licenses.md
- [X] Create a simple container-specific README.md in the same directory as the Dockerfile (i.e. spades/3.12.0/README.md)
- [x] Update GitHub actions workflow if needed
- [X] Any files required for building are located in the same directory as the Dockerfile (i.e. spades/3.12.0/my_spades_tests.sh)
- [x] Have successfully run the workflow "Test <program name> image" in your forked repository

erinyoung commented:
Dear @kapsakcj ,

This is currently stuck as a draft because spades isn't installing correctly. 

I'm probably missing something simple, but since you did spades' container, can you look over this one to see why it wouldn't install?

erinyoung commented:
This is currently stuck as a draft because spades isn't installing correctly. 

I'm probably missing something simple, but since you did spades' container, can you look over this one to see why it wouldn't install?

kapsakcj commented:
So I think ubuntu:jammy is an OS that is _too new_. I think SPAdes linux pre-compiled binaries are compiled on older OS's and thus older linux kernels.

https://github.com/ablab/spades#downloading-spades-linux-binaries This section in the docs mention:

> Note, that pre-build binaries do not work on new Linux kernels.

When I dropped down to ubuntu:focal as a base, I was able to get `spades.py --test` to run as expected.

I think downgrading to ubuntu:focal is the path of least resistance here. Shouldn't have to make any other changes other than changing the base image. Unless you want to go through the steps of compiling SPAdes from source. That should resolve it too and you could stick with ubuntu:jammy.

One other thing to note while testing - I could not get the test data to assemble properly. Kept running into spades errors while trying to do the hybrid assembly.

example errors:
```
spades.py -o /data/test/spades_assembly -k 25,49 --threads 4 --restart-from k25 -m 1024
Error: SPAdes failed to produce an assembly graph: /data/test/spades_assembly/K49/assembly_graph_with_scaffolds.gfa
```
another
```
spades.py -o /data/test/spades_assembly -k 25,49 --threads 4 --restart-from k25 -m 1024
Error: SPAdes failed to produce an assembly graph: /data/test/spades_assembly/K49/assembly_graph_with_scaffolds.gfa
```

Could you instead use test data that is supplied with unicycler code (and thus already in the docker image)? https://github.com/rrwick/Unicycler/tree/main/sample_data

This command should do it:
```
unicycler -1 /unicycler/sample_data/short_reads_1.fastq.gz -2 /unicycler/sample_data/short_reads_2.fastq.gz -l /unicycler/sample_data/long_reads_high_depth.fastq.gz --out sample_data_hybrid_asm_test
```

And to be extra thorough we could run unicycler's unit tests: https://github.com/rrwick/Unicycler/tree/main/test
```Dockerfile
RUN cd /unicycler && \
 python -m unittest
```

kapsakcj commented:
Spades is failing on the nf-core bacteroides fragilis dataset due to uneven kmer coverage:
```
 * 0:00:02.205     9M / 188M  ERROR   General                 (kmer_coverage_model.cpp   : 187)   Invalid kmer coverage histogram, make sure that the coverage is indeed uniform
 * system call for: "['/SPAdes-3.15.4-Linux/bin/spades-core', '/data/test/spades_assembly/K93/configs/config.info', '/data/test/spades_assembly/K93/configs/isolate_mode.info']" finished abnormally, OS return value: 21
```

If it's OK with you, I'd like to push some commits to get the test running properly

kapsakcj commented:
I made a few changes and was able to get all test running successfully. Now using the unicycler supplied test data instead of the nf-core bacteroides fragilis dataset, which was causing spades to fail.

Docker image builds successfully, and should do the same in the GH actions workflow.

I'm going to test locally with some Trycycler test data, I think it was called the "Great" dataset, which will be another good test.

kapsakcj commented:
Nevermind, going to try the other sets from Unicycler! Seems like trycycler only provides long reads and assemblies for testing.

https://github.com/rrwick/Unicycler/tree/main/sample_data

I'll try out the data from here

kapsakcj commented:
Shigella sonnei plasmid (synthetic data, also tested in test layer)
```
$ time unicycler -1 ../../testdata/unicycler-sample-data/shigella-sonnei-plasmids-synthetic-short-and-long-reads/short_reads_1.fastq.gz -2 ../../testdata/unicycler-sample-data/shigella-sonnei-plasmids-synthetic-short-and-long-reads/short_reads_2.fastq.gz -l ../../testdata/unicycler-sample-data/shigella-sonnei-plasmids-synthetic-short-and-long-reads/long_reads_high_depth.fastq.gz -o shigella-plasmid-hybrid

[...output redacted...]

Bridged assembly graph (2022-05-24 15:24:12)
    The assembly is now mostly finished and no more structural changes will be made. Ideally the assembly graph should now have one contig per replicon and no erroneous contigs (i.e. a
complete assembly). If there are more contigs, then the assembly is not complete.

Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/shigella-plasmid-hybrid/005_final_clean.gfa

Component   Segments   Links   Length    N50       Longest segment   Status
    total          3       3   229,880   215,774           215,774
        1          1       1   215,774   215,774           215,774   complete
        2          1       1     8,953     8,953             8,953   complete
        3          1       1     5,153     5,153             5,153   complete

Rotating completed replicons (2022-05-24 15:24:12)
    Any completed circular contigs (i.e. single contigs which have one link connecting end to start) can have their start position changed without altering the sequence. For consistency,
Unicycler now searches for a starting gene (dnaA or repA) in each such contig, and if one is found, the contig is rotated to start with that gene on the forward strand.

Segment   Length    Depth   Starting gene         Position   Strand    Identity   Coverage
      1   215,774   1.00x   UniRef90_D3H586        102,179   forward      94.4%      99.6%
      2     8,953   1.59x   UniRef90_A0A0H2QQL4      6,920   forward      96.5%     100.0%
      3     5,153   4.16x   none found

Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/shigella-plasmid-hybrid/006_rotated.gfa

Assembly complete (2022-05-24 15:25:03)
Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/shigella-plasmid-hybrid/assembly.gfa
Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/shigella-plasmid-hybrid/assembly.fasta

real    3m29.041s
user    15m42.971s
sys     0m11.879s
```

H. pylori:
```
$ time unicycler -1 ../../testdata/unicycler-sample-data/helicobacter-pylori-ilmn-pacbio/short_reads_1.fastq.gz -2 ../../testdata/unicycler-sample-data/helicobacter-pylori-ilmn-pacbio/short_reads_2.fastq.gz -l ../../testdata/unicycler-sample-data/helicobacter-pylori-ilmn-pacbio/long_reads_high_depth.fastq.gz -o helicobacter-pylori-hybid

[...output redacted...]

Bridged assembly graph (2022-05-24 16:53:23)
    The assembly is now mostly finished and no more structural changes will be made. Ideally the assembly graph should now have one contig per replicon and no erroneous contigs (i.e. a
complete assembly). If there are more contigs, then the assembly is not complete.

Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/helicobacter-pylori-hybid/006_final_clean.gfa

Component   Segments   Links   Length      N50         Longest segment   Status
        1         10      13   1,645,655   1,438,122         1,438,122   incomplete

Assembly complete (2022-05-24 16:53:23)
Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/helicobacter-pylori-hybid/assembly.gfa
Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/helicobacter-pylori-hybid/assembly.fasta

real    56m27.863s
user    377m0.911s
sys     3m57.009s
```

kapsakcj commented:
N. gonorrhoeae:
```
$ time unicycler -1 ../../testdata/unicycler-sample-data/ngonnorrhoeae-ilmn-and-pacbio-reads/short_reads_1.fastq.gz -2 ../../testdata/unicycler-sample-data/ngonnorrhoeae-ilmn-and-pacbio-reads/short_reads_2.fastq.gz -l ../../testdata/unicycler-sample-data/ngonnorrhoeae-ilmn-and-pacbio-reads/long_reads_high_depth.fastq.gz -o ngonnorrhoeae-hybrid

[...output redacted...]

Bridged assembly graph (2022-05-24 17:55:53)
    The assembly is now mostly finished and no more structural changes will be made. Ideally the assembly graph should now have one contig per replicon and no erroneous contigs (i.e. a
complete assembly). If there are more contigs, then the assembly is not complete.

Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/ngonnorrhoeae-hybrid/006_final_clean.gfa

Component   Segments   Links   Length      N50         Longest segment   Status
    total          2       2   2,216,669   2,212,516         2,212,516
        1          1       1   2,212,516   2,212,516         2,212,516   complete
        2          1       1       4,153       4,153             4,153   complete

Rotating completed replicons (2022-05-24 17:55:53)
    Any completed circular contigs (i.e. single contigs which have one link connecting end to start) can have their start position changed without altering the sequence. For consistency,
Unicycler now searches for a starting gene (dnaA or repA) in each such contig, and if one is found, the contig is rotated to start with that gene on the forward strand.

Segment   Length      Depth    Starting gene     Position    Strand    Identity   Coverage
      1   2,212,516    1.00x   UniRef90_Q9JXS7   1,850,435   forward      98.3%     100.0%
      2       4,153   12.42x   none found

Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/ngonnorrhoeae-hybrid/007_rotated.gfa

Assembly complete (2022-05-24 17:57:17)
Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/ngonnorrhoeae-hybrid/assembly.gfa
Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/ngonnorrhoeae-hybrid/assembly.fasta

real    49m12.273s
user    302m27.555s
sys     2m35.806s
```

Kept getting `Killed` OOM errors with S. pyogenes during this step: `Creating miniasm/Racon bridges`. The only way around it was to use `--no_miniasm` option. I believe this is a problem with the test data & Unicycler rather than the docker image.
```
$ time unicycler -1 ../../testdata/unicycler-sample-data/Spyogenes-ilmn-and-pacbio-reads/short_reads_1.fastq.gz -2 ../../testdata/unicycler-sample-data/Spyogenes-ilmn-and-pacbio-reads/short_reads_2.fastq.gz -l ../../testdata/unicycler-sample-data/Spyogenes-ilmn-and-pacbio-reads/long_reads_high_depth.fastq.gz -o spyogenes-hybrid --no_miniasm

[...output redacted...]

Bridged assembly graph (2022-05-25 00:19:42)
    The assembly is now mostly finished and no more structural changes will be made. Ideally the assembly graph should now have one contig per replicon and no
erroneous contigs (i.e. a complete assembly). If there are more contigs, then the assembly is not complete.

Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/spyogenes-hybrid/005_final_clean.gfa

Component   Segments   Links   Length      N50         Longest segment   Status
        1          6       7   1,741,359   1,735,358         1,735,358   incomplete

Assembly complete (2022-05-25 00:19:42)
Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/spyogenes-hybrid/assembly.gfa
Saving /data/outputs/2022-05-24-unicycler-0.5.0-tests/spyogenes-hybrid/assembly.fasta

real    20m49.310s
user    161m29.892s
sys     0m56.811s
```

Docker image works pretty well in my hands, would be great if you could test with some of your own data!

kapsakcj commented:
I'll delete `.github/workflows/test-unicycler.yml` and resolve the README.md conflict

erinyoung reviewed (commented):
Looks good to me! Since I created this PR, so I can't approve it.

I don't if the racon test is still needed, but it doesn't hurt to keep it in.

kapsakcj commented:
No worries about the delay, thanks for reviewing.

When I get a chance I will do one final lookover, give my approval and then deploy

erinyoung commented:
I forgot to include this before. Here's my attempt (everything worked great):
```
$ docker run -v $(pwd)/for_unicycler:/data -it unicycler:test
root@a3980d482628:/data# ls
1326944-1.fastq.gz     1326944-1_R2.fastq.gz  1326979-1_R1.fastq.gz  1326979-2.fastq.gz     1326979-2_R2.fastq.gz  1326979-3_R1.fastq.gz  1326983.fastq.gz     1326983_R2.fastq.gz
1326944-1_R1.fastq.gz  1326979-1.fastq.gz     1326979-1_R2.fastq.gz  1326979-2_R1.fastq.gz  1326979-3.fastq.gz     1326979-3_R2.fastq.gz  1326983_R1.fastq.gz
root@a3980d482628:/data# unicycler -l 1326944-1.fastq.gz -1 1326944-1_R1.fastq.gz -2 1326944-1_R2.fastq.gz -o unicycler_test/

Starting Unicycler (2022-06-23 17:03:11)
    Welcome to Unicycler, an assembly pipeline for bacterial genomes. Since you provided both short and long reads, Unicycler will perform a hybrid assembly. It will first use SPAdes to make a short-read assembly
graph, and then it will use various methods to scaffold that graph with the long reads.
    For more information, please see https://github.com/rrwick/Unicycler

Command: /usr/local/bin/unicycler -l 1326944-1.fastq.gz -1 1326944-1_R1.fastq.gz -2 1326944-1_R2.fastq.gz -o unicycler_test/

Unicycler version: v0.5.0
Using 8 threads

Making output directory:
  /data/unicycler_test

Dependencies:
  Program       Version   Status
  spades.py     3.15.4    good  
  racon         1.4.3     good  
  makeblastdb   2.12.0+   good  
  tblastn       2.12.0+   good  


Choosing k-mer range for assembly (2022-06-23 17:03:53)
    Unicycler chooses a k-mer range for SPAdes based on the length of the input reads. It uses a wide range of many k-mer sizes to maximise the chance of finding an ideal assembly.

SPAdes maximum k-mer: 127
Median read length: 149
K-mer range: 27, 53, 71, 87, 99, 111, 119, 127


SPAdes assemblies (2022-06-23 17:04:34)
    Unicycler now uses SPAdes to assemble the short reads. It scores the assembly graph for each k-mer using the number of contigs (fewer is better) and the number of dead ends (fewer is better). The score
function is 1/(c*(d+2)), where c is the contig count and d is the dead end count.

spades.py -o /data/unicycler_test/spades_assembly -k 27 --threads 8 --isolate -1 /data/1326944-1_R1.fastq.gz -2 /data/1326944-1_R2.fastq.gz -m 1024

spades.py -o /data/unicycler_test/spades_assembly -k 27,53 --threads 8 --restart-from k27 -m 1024

spades.py -o /data/unicycler_test/spades_assembly -k 27,53,71 --threads 8 --restart-from k53 -m 1024

spades.py -o /data/unicycler_test/spades_assembly -k 27,53,71,87 --threads 8 --restart-from k71 -m 1024

spades.py -o /data/unicycler_test/spades_assembly -k 27,53,71,87,99 --threads 8 --restart-from k87 -m 1024

spades.py -o /data/unicycler_test/spades_assembly -k 27,53,71,87,99,111 --threads 8 --restart-from k99 -m 1024

spades.py -o /data/unicycler_test/spades_assembly -k 27,53,71,87,99,111,119 --threads 8 --restart-from k111 -m 1024

spades.py -o /data/unicycler_test/spades_assembly -k 27,53,71,87,99,111,119,127 --threads 8 --restart-from k119 -m 1024

K-mer   Contigs   Dead ends   Score      
   27                         too complex
   53       477           7      2.33e-04
   71       363          17      1.45e-04
   87       220          11      3.50e-04
   99       181          11      4.25e-04
  111       164           7      6.78e-04
  119       160           8      6.25e-04
  127       153           6      8.17e-04 ← best

Read depth filter: removed 16 contigs totalling 4503 bp
Deleting /data/unicycler_test/spades_assembly/


Determining graph multiplicity (2022-06-23 17:35:08)
    Multiplicity is the number of times a sequence occurs in the underlying sequence. Single-copy contigs (those with a multiplicity of one, occurring only once in the underlying sequence) are particularly
useful.

Saving /data/unicycler_test/002_depth_filter.gfa


Cleaning graph (2022-06-23 17:35:08)
    Unicycler now performs various cleaning procedures on the graph to remove overlaps and simplify the graph structure. The end result is a graph ready for bridging.

Graph overlaps removed

Removed zero-length segments:
    105, 107, 110, 121, 132, 148

Removed zero-length segments:
    92, 100

Merged small segments:
    128, 133, 134, 135, 136, 137, 138, 139, 140, 141, 142, 143, 146, 147, 150, 151, 152

Saving /data/unicycler_test/003_overlaps_removed.gfa

    Unicycler now selects a set of anchor contigs from the single-copy contigs. These are the contigs which will be connected via bridges to form the final assembly.

40 anchor segments (3,879,763 bp) out of 128 total segments (3,913,577 bp)


Creating SPAdes contig bridges (2022-06-23 17:35:08)
    SPAdes uses paired-end information to perform repeat resolution (RR) and produce contigs from the assembly graph. SPAdes saves the graph paths corresponding to these contigs in the contigs.paths file. When
one of these paths contains two or more anchor contigs, Unicycler can create a bridge from the path.

                                                          Bridge
Start                    Path                    End     quality
    4               69 → -57 → 68                38         23.2
    5                    -74                     14         61.3
    8   -84 → -91 → 78 → -45 → -76 → 118 → -77   -29        14.6
    9   -84 → 90 → 78 → 47 → -76 → -119 → -77    29         17.3
   11    83 → 110 → -86 → 52 → 82 → 112 → 64     26         19.2
   12               87 → 49 → -75                2          23.0
   15               -68 → 56 → -69               24         22.5
   16                    -74                     5          62.7
   20               87 → 50 → -75                -36        22.8
   22    83 → 109 → -86 → 53 → 82 → -113 → 64    19         19.2
   26               -73 → -70 → 79               36         32.7
   31               -79 → 71 → 73                15         32.8


Creating loop unrolling bridges (2022-06-23 17:35:08)
    When a SPAdes contig path connects an anchor contig with the middle contig of a simple loop, Unicycler concludes that the sequences are contiguous (i.e. the loop is not a separate piece of DNA). It then uses
the read depth of the middle and repeat contigs to guess the number of times to traverse the loop and makes a bridge.

No loop unrolling bridges made


Loading reads (2022-06-23 17:35:08)
117,242 / 117,242 (100.0%) - 1,183,265,248 bp


Assembling contigs and long reads with miniasm (2022-06-23 17:35:52)
    Unicycler uses miniasm to construct a string graph assembly using both the short read contigs and the long reads. It will then use the resulting string graph to produce bridges between contigs. This method
requires decent coverage of long reads and therefore may not be fruitful if long reads are sparse. However, it does not rely on the short read assembly graph having good connectivity and is able to bridge an
assembly graph even when it contains many dead ends.
    Unicycler uses two types of "reads" as assembly input: anchor contigs from the short-read assembly and actual long reads which overlap two or more of these contigs. It then assembles them with miniasm.

Aligning long reads to graph using minimap

Saving to /data/unicycler_test/miniasm_assembly/01_assembly_reads.fastq:
  40 short-read contigs
  11,485 long reads

Finding overlaps with minimap... success
  2,321,742 overlaps

Assembling reads with miniasm... success
  185 segments, 185 links

Saving /data/unicycler_test/miniasm_assembly/11_branching_paths_removed.gfa
Merging segments into unitigs:
  3 circular unitigs
  total size = 3,986,643 bp
Saving /data/unicycler_test/miniasm_assembly/12_unitig_graph.gfa


Polishing miniasm assembly with Racon (2022-06-23 17:39:09)
    Unicycler now uses Racon to polish the miniasm assembly. It does multiple rounds of polishing to get the best consensus. Circular unitigs are rotated between rounds such that all parts (including the ends)
are polished well.

Saving to /data/unicycler_test/miniasm_assembly/racon_polish/polishing_reads.fastq:
  40 short-read contigs
  117,242 long reads

Polish       Assembly          Mapping
round            size          quality
begin       3,986,643        95,414.47
1           4,000,767       103,882.68

Best polish: /data/unicycler_test/miniasm_assembly/racon_polish/006_rotated.fasta
Saving /data/unicycler_test/miniasm_assembly/13_racon_polished.gfa
Saving /data/unicycler_test/004_long_read_assembly.gfa

    Contigs in the short-read assembly graph which end in dead ends may contain bogus sequence near the dead end. Unicycler therefore uses the read clipping values from the miniasm assembly to trim these dead
ends to only the parts which aligned well to long reads.

                       Trimmed   Trimmed             
            Previous      from      from        Final
Segment       length     start       end       length
23            54,338         -        65       54,273
37            19,565         -         4       19,561
3            284,727        51         -      284,676
9            151,539        13         -      151,526

    Unicycler now places the single copy contigs back into the unitig graph. This serves two purposes: a) it replaces long read assembly sequences (which may be error prone) with Illumina assembly sequence (which
is probably quite accurate), improving the assembly quality, and b) it defines inter-contig sequences for use in building bridges.

Searching for contigs using 5,000 bp of contig ends.

Contig   Result              Start pos   End pos   Strand
     1   found in unitig 1      825610   1213551        -
     2   found in unitig 1      125023    488630        +
     3   found in unitig 1     2951992   3236713        +
     4   found in unitig 1     2274913   2548103        -
     5   found in unitig 1     3324002   3595532        -
     6   found in unitig 1     1472757   1645572        -
     7   found in unitig 1     1231891   1403168        +
     8   found in unitig 1     2595767   2754127        +
     9   found in unitig 1     2800432   2951954        -
    10   found in unitig 1     3800405   3939082        +
    11   found in unitig 1     1646753   1758395        +
    12   found in unitig 1       14925    124157        +
    13   found in unitig 1      620509    711778        -
    14   found in unitig 1     3237894   3323854        -
    15   found in unitig 1     1943684   2024777        -
    16   found in unitig 1     3595680   3667219        -
    17   found in unitig 1      550382    619326        -
    18   found in unitig 1     2066358   2134558        -
    19   found in unitig 1     3681543   3743709        -
    20   found in unitig 1     1828471   1888685        -
    21   found in unitig 1     2145365   2204533        -
    22   found in unitig 1     3744904   3799281        -
    23   found in unitig 1     2204522   2258789        -
    24   found in unitig 1     1889865   1942849        -
    25   found in unitig 1      494192    544546        +
    26   found in unitig 1     1759591   1806625        +
    27   found in unitig 1     1426193   1471577        +
    28   found in unitig 1     2549284   2593245        +
    29   found in unitig 1     2755967   2798993        -
    30   found in unitig 1      748429    790776        +
    31   found in unitig 1     2025266   2065177        -
    32   found in unitig 1      -17271      9039        -
    33   found in unitig 1      717573    743625        +
    34   found in unitig 1      800054    824407        +
    35   found in unitig 1     1404349   1424990        +
    36   found in unitig 1     1807114   1827605        +
    37   found in unitig 2      -14014      5540        -
    38   found in unitig 1     2259969   2274078        -
    39   found in unitig 1     1219348   1230710        -
    88   not found                                       

Searching for contigs using 2,500 bp of contig ends.

Contig   Result      Start pos   End pos   Strand
    88   not found                               

Searching for contigs using 1,000 bp of contig ends.

Contig   Result      Start pos   End pos   Strand
    88   not found                               

Searching for contigs using 500 bp of contig ends.

Contig   Result      Start pos   End pos   Strand
    88   not found                               

Saving /data/unicycler_test/miniasm_assembly/14_contigs_placed.gfa


Creating miniasm/Racon bridges (2022-06-23 17:55:49)
    Now that the miniasm/Racon string graph is complete, Unicycler will use it to build bridges between anchor segments.

       Start → end     Best path                                 Quality
 1/39    -32 → 12      -104, 117, 105, -41, -97, 72               67.451
 2/39     -1 → -39     116, -92, 97, 41, -105, -117               62.736
 3/39    -39 → 7       -46, 103                                   89.805
 4/39      7 → 35      -46, 102                                   89.430
 5/39     35 → 27      -51, -127, -85, -120, 60, -121, 67         87.308
 6/39     27 → -6      -102, 46                                   88.267
 7/39     -6 → 11      -46, 102                                   87.746
 8/39     11 → 26      83, 110, -86, 52, 82, 112, 64              89.700
 9/39     26 → 36      -73, -70, 79                               95.216
10/39     36 → -20     75, -50, -87                               91.370
11/39    -20 → -24     -102, 46                                   78.419
12/39     12 → 2       87, 49, -75                                91.962
13/39    -24 → -15     69, -56, 68                                88.655
14/39    -15 → -31     -73, -71, 79                               95.425
15/39    -31 → -18     -46, 103                                   89.940
16/39    -18 → -21     -54, 55, 54, 40, -55, 54, 44, -55, 54      56.098
17/39    -23 → -38     -103, 46                                   89.765
18/39    -38 → -4      -68, 57, -69                               91.313
19/39     -4 → 28      -46, 103                                   86.272
20/39     28 → 8                                                  65.248
21/39      8 → -29     -84, -91, 78, -45, -76, 118, -77           82.507
22/39      2 → 25      105, -41                                   68.807
23/39    -29 → -9      77, 119, 76, -47, -78, -90, 84             87.975
24/39     -9 → 3                                                  97.346
25/39      3 → -14     -102, 46                                   88.238
26/39    -14 → -5      74                                         95.482
27/39     -5 → -16     74                                         97.776
28/39    -16 → -19     58, 122, 89, 101, 58, -124, 108, -99,      50.304
                       101, 58, -124, 89, 101, 58, 122, 89,             
                       101, 58, 122, 89, 101, 58, 122, 89, 101,         
                       58, 122, 89, 101, 58, -124, 108, 98, 65,         
                       100, 111, 107, 95, 62, -115, -81, 95,            
                       62, -115, -81, 95, 62, -114, -81, 95,            
                       62, -114, -81, 95, 62, -115, -81, -94,           
                       62, -115, -81, -94, 62, -115, -81, 95,           
                       62, -114, -81, -94, 62, -115, -81, -94,          
                       62, -115, -81, 95, 62, -114, -81, 95,            
                       62, -114, -48, 106, -96, -66, 100, 93,           
                       -96, -66, 80, 106, -96, 65, 100, 111,            
                       -61, -59, -63, -59                               
29/39    -19 → -22     -64, 113, -82, -53, 86, -109, -83          89.771
30/39    -22 → 10      -46                                        88.788
31/39     10 → -32     -41, -97, 72                               68.104
32/39    -37 → -37                                                98.957
33/39     25 → -17     -92, 97, 41, -105, -117, 104               67.319
34/39    -17 → -13     -46, 102                                   89.492
35/39    -13 → 33      116, -92, 97, 41, -105, -117               68.051
36/39     33 → 30      -102, 46, -42, -46, 102                    71.232
37/39     30 → 34      -43, -103, 46, 42, -46, 103, 43            58.430
38/39     34 → -1      -51, -126, -85, 123, 60, 125, 67           89.710
39/39    -21 → -23                                                99.852


Creating simple long read bridges (2022-06-23 17:56:20)
    Unicycler uses long read alignments (from minimap) to resolve simple repeat structures in the graph. This takes care of some "low-hanging fruit" of the graph simplification.

Aligning long reads to graph using minimap

    Two-way junctions are defined as cases where two graph contigs (A and B) join together (C) and then split apart again (D and E). This usually represents a simple 2-copy repeat, and there are two possible
options for its resolution: (A→C→D and B→C→E) or (A→C→E and B→C→D). Each read which spans such a junction gets to "vote" for option 1, option 2 or neither. Unicycler creates a bridge at each junction for the most
voted for option.

                                              Op. 1   Op. 2   Neither   Final    Bridge
Junction       Option 1         Option 2      votes   votes     votes     op.   quality
      74   -14 → 74 → -16,   -14 → 74 → -5,       0     478         3       2      98.6
             -5 → 74 → -5    -5 → 74 → -16                                             

    Simple loops are parts of the graph where two contigs (A and B) are connected via a repeat (C) which loops back to itself (via D). It is possible to traverse the loop zero times (A→C→B), one time (A→C→D→C→B),
two times (A→C→D→C→D→C→B), etc. Long reads which span the loop inform which is the correct number of times through. In this step, such reads are found and each is aligned against alternative loop counts. A reads
casts its "vote" for the loop count it agrees best with, and Unicycler creates a bridge using the most voted for count.

No suitable simple loops present


Determining low score threshold (2022-06-23 17:56:56)
    Before conducting semi-global alignment of the long reads to the assembly graph, Unicycler must determine a minimum alignment score threshold such that nonsense alignments are excluded. To choose a threshold
automatically, it examines alignments between random sequences and selects a score a few standard deviations above the mean.

Automatically choosing a threshold using random alignment scores.

Random alignment mean score: 61.66
         standard deviation:  1.31
        Low score threshold: 61.66 + (7 x 1.31) = 70.86


Aligning reads (2022-06-23 17:57:17)
117,242 / 117,242 (100.0%)


Read alignment summary (2022-06-23 20:41:17)
Total read count:              117,242
Fully aligned reads:           107,968
Partially aligned reads:         8,527
Unaligned reads:                   747
Total bases aligned:     1,172,320,112 bp
Mean alignment identity:            94.8%

Deleting /data/unicycler_test/read_alignment/


Building long read bridges (2022-06-23 20:41:18)
    Unicycler uses the long read alignments to produce bridges between anchor segments. These bridges can be formed using as few as one long read, giving
Unicycler the ability to bridge the graph even when long-read depth is low.

       Start → end     Best path                                 Quality
 1/59     39 → 1       117, 105, -41, -97, 92, -116               76.116
 2/59    -12 → 32      -72, 97, 41, -105, -117, 104               82.911
 3/59     32 → -10     -72, 97, 41                                83.277
 4/59    -13 → 33      116, -92, 97, 41, -105, -117               83.135
 5/59      5 → -3      -74, 14, -46, 102                          22.539
 6/59     34 → 28                                                  0.708
 7/59     25 → -17     -92, 97, 41, -105, -117, 104               82.197
 8/59     15 → 20      -68, 56, -69, 24, -46, 102                 14.256
 9/59     30 → 34      -43, -103, 46, 42, -46, 103, 43            81.737
10/59     33 → 30      -102, 46, -42, -46, 102                    82.333
11/59     28 → 8                                                  28.153
12/59      2 → 25      105, -41                                   83.055
13/59    -13 → 30      116, -92, 97, 41, -105, -117, 33, -102,    25.292
                       46, -42, -46, 102                                
14/59     29 → -8      77, -118, 76, 45, -78, 91, 84              80.666
15/59     -4 → 21                                                  2.778
16/59     21 → 18      -54, 55, -44, -54, 55, -40, -54, -55, 54   83.438
17/59      9 → 29      -84, 90, 78, 47, -76, -119, -77            83.358
18/59     35 → 27      -51, -127, -85, -120, 60, -121, 67         80.617
19/59     30 → -1      -43, -103, 46, 42, -46, 103, 43, 34,       24.214
                       -51, -126, -85, 123, 60, 125, 67                 
20/59     14 → -3      -46, 102                                   81.097
21/59     13 → 17      -102, 46                                   83.181
22/59     26 → -20     -73, -70, 79, 36, 75, -50, -87             28.147
23/59     24 → 20      -46, 102                                   71.554
24/59     34 → -1      -51, -126, -85, 123, 60, 125, 67           83.299
25/59     11 → 26      83, 110, -86, 52, 82, 112, 64              83.439
26/59     27 → -6      -102, 46                                   80.843
27/59     -7 → 39      -103, 46                                   81.274
28/59     -6 → 11      -46, 102                                   80.656
29/59     22 → 19      83, 109, -86, 53, 82, -113, 64             83.455
30/59     18 → 31      -103, 46                                   82.930
31/59     -4 → 28      -46, 103                                   78.765
32/59      7 → 27      -46, 102, 35, -51, -127, -85, -120, 60,    33.176
                       -121, 67                                         
33/59    -10 → 22      46                                         82.167
34/59     38 → 23      -46, 103                                   81.387
35/59     36 → -20     75, -50, -87                               82.090
36/59      7 → 35      -46, 102                                   81.542
37/59     26 → 30                                                  0.966
38/59    -30 → 34      -102, 46, 42, -46, 103, 43                  9.987
39/59     33 → -30     -102, 46, -42, -46, 103, 43                 6.793
40/59     15 → 24      -68, 56, -69                               80.086
41/59     28 → 34      -102, 46, 42, -46, 103, 43                  1.872
42/59     12 → 2       87, 49, -75                                83.541
43/59      4 → 38      69, -57, 68                                81.599
44/59     31 → 15      -79, 71, 73                                83.498
45/59     38 → -28     -46, 102                                    8.918
46/59     -4 → 8       -46, 102                                    6.774
47/59      7 → 10      -46                                         3.005
48/59     26 → 36      -73, -70, 79                               82.841
49/59     17 → 25      -104, 117, 105, -41                        12.277
50/59     37 → 37                                                 43.121
51/59     16 → 5       -74                                        78.558
52/59      5 → 14      -74                                        81.196
53/59     23 → 21                                                 30.100
54/59     -3 → 9                                                  96.159
55/59     25 → -2      -92, 97, 41, -105                          14.731
56/59     19 → 16      59, 63, 59, 61, -111, -100, -65, 96,       82.203
                       -106, -80, 66, 96, -93, -100, 66, 96,            
                       -106, 48, 114, -62, -95, 81, 114, -62,           
                       -95, 81, 115, -62, 94, 81, 115, -62, 94,         
                       81, 114, -62, -95, 81, 115, -62, 94, 81,         
                       115, -62, 94, 81, 115, -62, -95, 81,             
                       114, -62, -95, 81, 114, -62, -95, 81,            
                       115, -62, -95, 81, 115, -62, -95, -107,          
                       -111, -100, -65, -98, -108, 124, -58,            
                       -101, -89, -122, -58, -101, -89, -122,           
                       -58, -101, -89, -122, -58, -101, -89,            
                       -122, -58, -101, -89, 124, -58, -101,            
                       99, -108, 124, -58, -101, -89, -122, -58         
57/59      4 → 23      69, -57, 68, 38, -46, 103                  60.388
58/59     -1 → 7       116, -92, 97, 41, -105, -117, -39, -46,    54.333
                       103                                              
59/59      9 → -8      -84, 90, 78, 47, -76, -119, -77, 29, 77,   12.695
                       -118, 76, 45, -78, 91, 84                        


Applying bridges (2022-06-23 20:44:08)
    Unicycler now applies to the graph in decreasing order of quality. This ensures that when multiple, contradictory bridges exist, the most supported option
is used.

Bridge type        Start → end   Path                                       Quality
miniasm              -21 → -23                                               99.852
miniasm              -37 → -37                                               98.957
simple long read      -5 → -16   74                                          98.592
miniasm               -9 → 3                                                 97.346
simple long read     -14 → -5    74                                          96.278
miniasm              -15 → -31   -73, -71, 79                                95.425
miniasm               26 → 36    -73, -70, 79                                95.216
miniasm               12 → 2     87, 49, -75                                 91.962
miniasm               36 → -20   75, -50, -87                                91.370
miniasm              -38 → -4    -68, 57, -69                                91.313
miniasm              -31 → -18   -46, 103                                    89.940
miniasm              -39 → 7     -46, 103                                    89.805
miniasm              -19 → -22   -64, 113, -82, -53, 86, -109, -83           89.771
miniasm              -23 → -38   -103, 46                                    89.765
miniasm               34 → -1    -51, -126, -85, 123, 60, 125, 67            89.710
miniasm               11 → 26    83, 110, -86, 52, 82, 112, 64               89.700
miniasm              -17 → -13   -46, 102                                    89.492
miniasm                7 → 35    -46, 102                                    89.430
miniasm              -22 → 10    -46                                         88.788
miniasm              -24 → -15   69, -56, 68                                 88.655
miniasm               27 → -6    -102, 46                                    88.267
miniasm                3 → -14   -102, 46                                    88.238
miniasm              -29 → -9    77, 119, 76, -47, -78, -90, 84              87.975
miniasm               -6 → 11    -46, 102                                    87.746
miniasm               35 → 27    -51, -127, -85, -120, 60, -121, 67          87.308
miniasm               -4 → 28    -46, 103                                    86.272
long read             21 → 18    -54, 55, -44, -54, 55, -40, -54, -55, 54    83.438
long read             32 → -10   -72, 97, 41                                 83.277
long read            -13 → 33    116, -92, 97, 41, -105, -117                83.135
long read              2 → 25    105, -41                                    83.055
long read            -12 → 32    -72, 97, 41, -105, -117, 104                82.911
miniasm                8 → -29   -84, -91, 78, -45, -76, 118, -77            82.507
long read             33 → 30    -102, 46, -42, -46, 102                     82.333
long read             19 → 16    59, 63, 59, 61, -111, -100, -65, 96,        82.203
                                 -106, -80, 66, 96, -93, -100, 66, 96,             
                                 -106, 48, 114, -62, -95, 81, 114, -62,            
                                 -95, 81, 115, -62, 94, 81, 115, -62, 94,          
                                 81, 114, -62, -95, 81, 115, -62, 94, 81,          
                                 115, -62, 94, 81, 115, -62, -95, 81,              
                                 114, -62, -95, 81, 114, -62, -95, 81,             
                                 115, -62, -95, 81, 115, -62, -95, -107,           
                                 -111, -100, -65, -98, -108, 124, -58,             
                                 -101, -89, -122, -58, -101, -89, -122,            
                                 -58, -101, -89, -122, -58, -101, -89,             
                                 -122, -58, -101, -89, 124, -58, -101,             
                                 99, -108, 124, -58, -101, -89, -122, -58          
long read             25 → -17   -92, 97, 41, -105, -117, 104                82.197
long read             30 → 34    -43, -103, 46, 42, -46, 103, 43             81.737
miniasm              -20 → -24   -102, 46                                    78.419
long read             39 → 1     117, 105, -41, -97, 92, -116                76.116
miniasm               28 → 8                                                 65.248

Saving /data/unicycler_test/005_bridges_applied.gfa


Bridged assembly graph (2022-06-23 20:44:09)
    The assembly is now mostly finished and no more structural changes will be made. Ideally the assembly graph should now have one contig per replicon and no
erroneous contigs (i.e. a complete assembly). If there are more contigs, then the assembly is not complete.

Saving /data/unicycler_test/006_final_clean.gfa

Component   Segments   Links   Length      N50         Longest segment   Status  
    total          2       2   3,981,663   3,961,996         3,961,996           
        1          1       1   3,961,996   3,961,996         3,961,996   complete
        2          1       1      19,667      19,667            19,667   complete


Rotating completed replicons (2022-06-23 20:44:09)
    Any completed circular contigs (i.e. single contigs which have one link connecting end to start) can have their start position changed without altering
the sequence. For consistency, Unicycler now searches for a starting gene (dnaA or repA) in each such contig, and if one is found, the contig is rotated to
start with that gene on the forward strand.

Segment   Length      Depth   Starting gene     Position   Strand    Identity   Coverage
      1   3,961,996   1.00x   UniRef90_Q6FG21    686,776   reverse      91.2%     100.2%
      2      19,667   3.53x   none found                                                

Saving /data/unicycler_test/007_rotated.gfa


Assembly complete (2022-06-23 20:45:22)
Saving /data/unicycler_test/assembly.gfa
Saving /data/unicycler_test/assembly.fasta

```
